src/gui/qgsowssourcewidget.cpp

```diff
@@ -0,0 +1,176 @@
+
+  // 'Prefer network' is the default noted in the combobox's tool tip
+  mCacheComboBox->setCurrentIndex( mCacheComboBox->findData( QNetworkRequest::PreferNetwork ) );
+
+  if ( providerKey == QStringLiteral( "wcs" ) )
+    mWMSGroupBox->hide();
+
+}
+
+void QgsOWSSourceWidget::clearCrs()
+{
+  mCRSLabel->setText( tr( "Coordinate Reference System" ) + ':' );
+  mSelectedCRSLabel->clear();
+  mChangeCRSButton->setEnabled( false );
+}
+
+void QgsOWSSourceWidget::clearFormats()
+{
+  mFormatComboBox->clear();
+  mFormatComboBox->setEnabled( false );
+}
+
+void QgsOWSSourceWidget::clearTimes()
+{
+  mTimeComboBox->clear();
+  mTimeComboBox->setEnabled( false );
+}
+
+void QgsOWSSourceWidget::populateTimes()
+{
+  mTimeComboBox->clear();
+  mTimeComboBox->insertItems( 0, times() );
+  mTimeComboBox->setEnabled( !times().isEmpty() );
+}
+
+QString QgsOWSSourceWidget::selectedTimeText()
+{
+  return mTimeComboBox->currentText();
+}
+
+void QgsOWSSourceWidget::hideInputWidgets()
+{
+  mTimeWidget->hide();
+  mFormatWidget->hide();
+  mCRSWidget->hide();
+  mCacheWidget->hide();
+}
+
+void QgsOWSSourceWidget::insertItemFormat( int index, const QString &label )
+{
+  mFormatComboBox->insertItem( index, label );
+}
+
+void QgsOWSSourceWidget::setFormatCurrentIndex( int index )
+{
+  mFormatComboBox->setCurrentIndex( index );
+}
+
+int QgsOWSSourceWidget::formatCurrentIndex()
+{
+  return mFormatComboBox->currentIndex();
+}
+
+void QgsOWSSourceWidget::setFormatEnabled( bool enabled )
+{
+  mFormatComboBox->setEnabled( enabled );
+}
+
+void QgsOWSSourceWidget::mChangeCRSButton_clicked()
+{
+  emit changeCRSButtonClicked();
+}
+
+void QgsOWSSourceWidget::setSelectedCRSLabel( const QString &label )
+{
+  mCRSLabel->setText( label );
+}
+
+void QgsOWSSourceWidget::setCRSLabel( const QString &label )
+{
+  mSelectedCRSLabel->setText( label );
+}
+
+void QgsOWSSourceWidget::setChangeCRSButtonEnabled( bool enabled )
+{
+  mChangeCRSButton->setEnabled( enabled );
+}
+
+QVariant QgsOWSSourceWidget::cacheData()
+{
+  return mCacheComboBox->currentData();
+}
+
+void QgsOWSSourceWidget::prepareExtent( QgsMapCanvas *mapCanvas )
+{
+  QgsCoordinateReferenceSystem crs = QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:4326" ) );
+  mSpatialExtentBox->setOutputCrs( crs );
+  if ( !mapCanvas )
+    return;
+  QgsCoordinateReferenceSystem destinationCrs = mapCanvas->mapSettings().destinationCrs();
+  mSpatialExtentBox->setCurrentExtent( destinationCrs.bounds(), destinationCrs );
+  mSpatialExtentBox->setOutputExtentFromCurrent();
+  mSpatialExtentBox->setMapCanvas( mapCanvas );
+}
+
+bool QgsOWSSourceWidget::extentChecked()
+{
+  return mSpatialExtentBox->isChecked();
+}
+
+QgsRectangle QgsOWSSourceWidget::outputExtent()
+{
+  return mSpatialExtentBox->outputExtent();
+}
+
+void QgsOWSSourceWidget::setSourceUri( const QString &uri )
+{
+  mSourceParts = QgsProviderRegistry::instance()->decodeUri( mProviderKey, uri );
```

Comment:
This would also need to set the widget state to match the URI, e.g. setting the current value of the output extent widget, crs, times widgets, cache combo value, tile widgets, etc

Basically the round trip from widget::setSourceUri() -> sourceUri() must be lossless (unless a user manually makes changes in the widget!)

Reply:
ok @nyalldawson working on adding this.